src/main/java/org/spongepowered/api/item/recipe/smelting/SmeltingRecipe.java

```diff
@@ -0,0 +1,153 @@
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item.recipe.smelting;
+
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.item.inventory.ItemStackSnapshot;
+import org.spongepowered.api.item.recipe.Recipe;
+import org.spongepowered.api.util.ResettableBuilder;
+
+import javax.annotation.Nullable;
+import java.util.Optional;
+
+/**
+ * A general interface for furnace recipes. You can implement it manually to
+ * suit your creative needs, or you can simply use the
+ * {@link SmeltingRecipe.Builder}.
+ */
+public interface SmeltingRecipe extends Recipe {
+    /**
+     * Builds a simple furnace recipe. Note, that you can implement the
+     * {@link SmeltingRecipe} manually, too.
+     */
+    static Builder builder() {
+        return Sponge.getRegistry().createBuilder(Builder.class);
+    }
+
+    /**
+     * An exemplary {@link ItemStackSnapshot}, which will always make
+     * {@link #isValid(ItemStackSnapshot)} return {@code true}.
+     *
+     * @return The {@link ItemStackSnapshot} as explained above
+     */
+    ItemStackSnapshot getExemplaryIngredient();
+
+    /**
+     * Checks if the given {@link ItemStackSnapshot} fits the required
+     * constraints to craft this {@link SmeltingRecipe}.
+     *
+     * @param ingredient The ingredient to check against
+     * @return Whether this {@param ingredient} can be used to craft the result
+     */
+    boolean isValid(ItemStackSnapshot ingredient);
+
+    /**
+     * This method should be used instead of the {@link #getExemplaryResult()}
+     * method, as it customizes the result further depending on the specified
+     * {@param ingredient} {@link ItemStackSnapshot}. It is advised to use
+     * the output of {@link #getExemplaryResult()}, modify it accordingly,
+     * and {@code return} it.
+     *
+     * @param ingredient The {@link ItemStackSnapshot} currently being smelted
+     * @return The result of smelting the {@param ingredient} {@link ItemStack},
+     *         if the {@param ingredient} is valid
+     */
+    Optional<ItemStack> getResult(ItemStackSnapshot ingredient);
+
+    /**
+     * Returns the amount of experience released after completing this recipe.
+     *
+     * @param ingredient The ingredient being smelted
+     * @return The amount of experience released after completing this recipe,
+     *         if the {@param ingredient} is valid
+     */
+    Optional<Double> getExperience(ItemStackSnapshot ingredient);
+
+    /**
+     * Builds a simple furnace recipe
+     */
+    interface Builder extends ResettableBuilder<SmeltingRecipe, Builder> {
+        /**
+         * Changes the result and returns this builder. The result is the
+         * {@link ItemStack} created when the recipe is fulfilled.
+         *
+         * @param result The output of this recipe
+         * @return This builder, for chaining
+         */
+        Builder result(@Nullable ItemStackSnapshot result);
```

Comment:
What are the implications of setting null here?
IMO setting it to null is useless since you have to set it to something none null anyway.